### ISSUE TITLE
[Feature request] Improving rescheduling

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs4anki-helper#175]
I think that most users see a huge backlog after rescheduling their cards when switching from Anki to FSRS or from FSRS v3 to FSRS v4. I think we need an additional load balancing feature when rescheduling a large (say, >1000 cards) amount of cards to spread the reviews across several days or a week so that switching from one algorithm to a different one is less "painful" in the sense that the user doesn't have to immediately do a lot of reviews.
Also, [someone on Reddit mentioned a problem with rescheduling and a large backlog](https://www.reddit.com/r/Anki/comments/15ggz5d/rescheduling_with_fsrs_continues_to_result_in/), but it seems that they didn't make a github issue.

_Originally posted by @Expertium in https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1666849832_

**Comment by galantra** [comment]
Isn't the _Postpone_ function made for this kind of use case?

**Comment by Expertium** [comment]
Yes, but I wonder if there is a better way of doing this, without using Postpone.

**Comment by Expertium** [comment]
By the way, I noticed that after rescheduling I get cards with R much higher than my requested R, which is 0.8 for all my decks.
![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/83031600/d7957350-9045-4d4e-8048-cfb457fcd5fb)

**Comment by user1823** [comment]
@Expertium, please re-read https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1666862580.

In my opinion, Sherlock would also agree that you should create new issues for things that are not directly related to the previous comments.

Possible causes for the issue you mentioned include
- a misconfiguration (in the scheduler code),
- you forgot to reschedule, or
- a bug.

But, none of these is related to the current issue.

**Comment by galantra** [comment]
To list the options:

1. Without FSRS4Anki:
	2. Set review sort order to “by overdueness”, 
	3. optionally set the deck option _Maximum reviews/day_,
	4. then reduce the backlog little by little, reviewing however many cards you can per day.   
2. Without the Postpone feature:
	1. Temporarily set the requested retention to a lower value, 
	2. then reschedule.  
3. Manual postpone.
4. Automatic Postpone after rescheduling, until a certain number of cards is reached.  
	- This would yet need to be implemented.  
	- It is similar to the deck setting _Maximum reviews/day_. Therefore, it should just use the number set in _Maximum reviews/day_ in the (rare) case that the user has set it. 
	- Otherwise, either the developers would need to set either 
		- a hard maximum and immediately postpone the excess
		- or a threshold where the Helper will ask the user whether and how many cards they want to Postpone: 
			> FSRS4Anki Helper has detected that rescheduling results in a large backlog of [n] cards. Would you like to postpone some of them?
			> Postpone |___| cards
			>                  |\_\_\_OK\_\_\_| |\_Cancel\_|

**Comment by Expertium** [comment]
> Automatic Postpone after rescheduling, until a certain number of cards is reached.
> * This would yet need to be implemented.
> * It is similar to the deck setting _Maximum reviews/day_. Therefore, it should just use the number set in _Maximum reviews/day_ in the (rare) case that the user has set it.
> * Otherwise, the developers would need to set either
>   
>   * a hard maximum and immediately postpone the excess
>   * or a threshold where the Helper will ask the user whether and how many cards they want to Postpone:
>     > FSRS4Anki Helper has detected that rescheduling results in a large backlog of [n] cards. This can happen, especially when using the rescheduler for the first time.
>     > Would you like to [postpone](https://github.com/open-spaced-repetition/fsrs4anki-helper#advancepostpone) some of the due cards to later dates?
>     > Postpone |___| cards
>     > |___OK___| |_Cancel_|

@L-M-Sherlock I support this idea. As for what is considered "large", how about burden (**after** rescheduling)*2? If the backlog is greater than that, then this message should appear.

**Comment by user1823** [comment]
In my opinion, the better way to handle backlogs is to work it off little by little, reviewing as many cards as possible per day. So, I don't use the postpone function for dealing with backlogs.

The reason is that postpone increases the interval of the card by a specific multiplication factor, which may even be greater than a month for cards with long intervals. But, when I clear backlogs by doing as many cards as possible per day, I am usually able to clear all of it within 10 days. So, using the postpone function would mean that clearing the backlog _completely_ would take longer as compared to not using the postpone function.

But, of course, this is my way of thinking about the problem. Others may or may not agree with it.

**Comment by L-M-Sherlock** [comment]
I prefer the method proposed by @user1823. `Postpone` just sweeps the problem under the rug. But I also agree that backlog will induce mental pressure. I think if we can set the default sorting of Anki to "Relative Overdueness", the problem could be solved easily via collaborating with setting the "Maximum reviews/day".

**Comment by Expertium** [comment]
So do you plan to implement a new setting, "Maximum reviews/day", into the scheduler?

**Comment by user1823** [comment]
> So do you plan to implement a new setting, "Maximum reviews/day", into the scheduler?

This option is already there in the Anki deck options.

**Comment by Expertium** [comment]
Yes, but I was asking whether Sherlock is planning to add it to FSRS in a way that overrides Anki's built-in setting, just like FSRS's max interval overrides Anki's max interval.

**Comment by user1823** [comment]
I think that Sherlock means to say that we can advise users to set the review sort order to "Relative Overdueness" and adjust the "Maximum reviews/day" in the Deck Options when faced with a backlog.

**Comment by Expertium** [comment]
I think we may need a new feature similar to Postpone, but targeting a specific number of reviews/day. The problem is that it's impossible to make it in a way that doesn't ruin the optimal scheduling. Changing how cards are spread in time will inevitably produce deviations from optimality. That being said, a new Postpone (whatever you want to call it) that spreads cards in a way that achieves a certain number of reviews/day might be the best way to deal with the backlog. Perhaps it's time we implement per-deck priority. Then the new feature can use not only predicted R, but also the priority of the deck as specified by the user, so that lower priority material is alloweed to deviate from optimal scheduling more.
@L-M-Sherlock here is my idea: call the new feature "Dilute" (yes, stolen from SuperMemo again). Unlike Postpone, which only deals with due cards, Dilute will affect both due and undue cards. It should change intervals in a way that achieves a user-specified value of Burden, or lower. For example, if user specifies 100, then this feature should rearrange reviews in such a way that Burden will drop to 100 or lower. Though unfortunately I can't describe a precise step-by-step algorithm for doing so. Of course, it has to be done in a way that produces the least deviation from the optimal scheduling.
The problem is that abusing this feature can be even mroe detrimental than abusing Postpone, since it will affect both due and and undue cards, meaning that it will lead to a larger deviation from optimality. And if the user specifies a very low number (say, his burden is normally 100, but he specifies 1), it can completely break the algorithm.

**Comment by poteau** [comment]
> By the way, I noticed that after rescheduling I get cards with R much higher than my requested R, which is 0.8 for all my decks. 

That R value is retrievability for the specific card, not your requested retention.

**Comment by Expertium** [comment]
When rescheduling occurs, cards with retrievability significantly higher than requested retention shouldn't be due. Sure, a little bit of it can happen due to siblings dispersion and load balancing, but not to the point where I have like 100 cards with retrievability being much higher than requested retention. For example, my requested retention is 0.8, and if I got 5 cars with R=80-85%, that's fine. Getting 100 cards with R>90% when requested R is 0.8 is definitely not intended behaviour.

**Comment by user1823** [comment]
Just a reminder to do this: https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/175#issuecomment-1675936970

But, there are some new points to consider. So, it is not recommended to start working on this until these points are discussed.

As we saw in https://www.reddit.com/r/Anki/comments/16d56xp/spits_coffee/, the burden can also be quite large.

So, we probably need a different cutoff. Probably, 1.75 × average daily review load in the past week.

In the linked case, the problem was the retention i.e., the user had a very low retention before switching to FSRS and the target retention was high.

So, probably, if the the above cutoff (1.75 × average review load) is exceeded, we can compare the true retention in the past week with the target retention. If the target retention is more than 10% higher than the past true retention, we can warn the user.

**Comment by L-M-Sherlock** [comment]
What's the state of this issue?

**Comment by user1823** [comment]
This feature is now not as important as it once was because we advise against rescheduling all cards on switching to FSRS.

However, it would still be a "good to have" feature.

**Comment by L-M-Sherlock** [comment]
I think it is a "good to have" feature in Anki because the native feature is more popular than the add-on.

**Comment by user1823** [comment]
> I think it is a "good to have" feature in Anki

I think that you are right. So, maybe we should add a feature request on Anki Forums.

**Comment by L-M-Sherlock** [comment]
What's the status of this issue? Can I close it?
